### PR TITLE
Separate Go Unit GHA into individual workflows

### PR DESCRIPTION
**Pull request by ericholguin**
This PR separates all of the Go unit tests from the go.unit.tests.yml workflow into their own individual workflow files, this allows us to have more granular GHA badges in the ReadMe as well as only running the specific unit test when the associated files are changed.

Few extra additions:
- Run ALL of the cache-config unit tests not just t3c-generate
- Added the v5 test files needed for the CiaB v5 TO integration tests to actually run.

<hr/>

## Which Traffic Control components are affected by this PR?

- CDN in a Box
- GitHub Actions

## What is the best way to verify this PR?

Make sure GHA are successful

## PR submission checklist
- [ ] This PR has tests 
- [ ] This PR has documentation 
- [ ] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: infrastructure/cdn-in-a-box/traffic_ops_integration_test/Dockerfile [hunk 1]
```diff
@@ -46,7 +46,7 @@ WORKDIR /go/src/github.com/apache/trafficcontrol/traffic_ops/testing/api
 RUN go mod vendor -v
 
 RUN set -o errexit -o xtrace;\
-    for api_version in v3 v4; do\
+    for api_version in v3 v4 v5; do\
```
Comment (zrhoffman) [inline infrastructure/cdn-in-a-box/traffic_ops_integration_test/Dockerfile:49]: `v3 v4 v5` can be `v{3..5}` like https://github.com/apache/trafficcontrol/blob/e9616117a52277f6d28bfee12873496d68e1ecae/infrastructure/cdn-in-a-box/traffic_ops_integration_test/run.sh#L45

File: infrastructure/cdn-in-a-box/traffic_ops_integration_test/Dockerfile [hunk 2]
```diff
@@ -71,9 +71,11 @@ COPY ./infrastructure/cdn-in-a-box/traffic_ops/to-access.sh /opt/integration/app
 COPY ./infrastructure/cdn-in-a-box/traffic_ops_integration_test/config.sh /opt/integration/app/
 COPY ./traffic_ops/testing/api/v3/tc-fixtures.json /opt/integration/app/tc-fixtures-v3.json
 COPY ./traffic_ops/testing/api/v4/tc-fixtures.json /opt/integration/app/tc-fixtures-v4.json
+COPY ./traffic_ops/testing/api/v5/tc-fixtures.json /opt/integration/app/tc-fixtures-v5.json
 COPY --from=integration-builder \
     /go/src/github.com/apache/trafficcontrol/traffic_ops/testing/api/traffic_ops_v3_integration_test \
     /go/src/github.com/apache/trafficcontrol/traffic_ops/testing/api/traffic_ops_v4_integration_test \
+    /go/src/github.com/apache/trafficcontrol/traffic_ops/testing/api/traffic_ops_v5_integration_test \
```
Comment (zrhoffman) [inline infrastructure/cdn-in-a-box/traffic_ops_integration_test/Dockerfile:78]: Good catch, that's been a lot of missing test coverage not running the v5 API tests when running from CiaB until now

**Review by zrhoffman — COMMENTED** [review]
What is the goal of splitting the Go unit tests into components?

If the goal is making it easier to identify which components are failing, I'd prefer highlighting which line of code a test failure occurred on using [golang-test-annotations](https://github.com/marketplace/actions/golang-test-annotations) or something similar.

Also, we shouldn't only run *some* unit tests on a given PR, we should run all unit tests every time, which we can do because unit tests are almost always fast.

Since we're running all Go unit tests on every Go-modifying PR, splitting the Go unit tests by component means occupying 6 runners for each such PR instead of 1, which is a lot of increased overhead, considering the shared Apache org's GHA runner pool, which would make sense if there was a corresponding very large benefit.

If you feel strongly that we would still benefit from splitting them up by component, though, that could be done more succinctly using a GHA [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) strategy. That way, we would still only need 1 workflow, even though it would still show up 6 times in the list of GHAs run.

**Review by ocket8888 — CHANGES_REQUESTED** [review]
One thing Zach didn't mention: according to what I'm seeing here, the lib tests only run if code in the lib changes. But that library is linked to from, I think, every other Go component. So when lib changes, we need to run ALL of those tests, not just the lib tests. Also, things that depend on the Go client need to be re-tested when the Go client changes. I'm actually not sure that we're doing that today - but if not it's a gap in testing that should be closed.

**Comment by ericholguin** [comment]
> What is the goal of splitting the Go unit tests into components?

Since we'd like to run all the unit tests regardless of files modified, the only goal would be to have separated badges, GitHub actions only have one badge per workflow, hence they all need a separate workflow. The value gain is that you'll be able to see which specific unit test failed at a glance without needing to click into the whole workflow. Also it would allow for Codecov badges to be added alongside to each of the unit workflow badges, showing the coverage for each unit tests and not the coverage as a whole.

> Also, we shouldn't only run _some_ unit tests on a given PR, we should run all unit tests every time, which we can do because unit tests are almost always fast.

Agreed, will fix that

> If you feel strongly that we would still benefit from splitting them up by component, though, that could be done more succinctly using a GHA [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) strategy. That way, we would still only need 1 workflow, even though it would still show up 6 times in the list of GHAs run.

Again, the badge workflow issue, will explore if there is an alternative

**Comment by ericholguin** [comment]
> Since we're running all Go unit tests on every Go-modifying PR, splitting the Go unit tests by component means occupying 6 runners for each such PR instead of 1, which is a lot of increased overhead, considering the shared Apache org's GHA runner pool, which would make sense if there was a corresponding very large benefit.

I think it already uses a runner per job, so we are already using 6 runners: https://docs.github.com/en/actions/using-workflows/about-workflows <-- See 2 under Workflow basics

**Comment by zrhoffman** [comment]
> > Since we're running all Go unit tests on every Go-modifying PR, splitting the Go unit tests by component means occupying 6 runners for each such PR instead of 1, which is a lot of increased overhead, considering the shared Apache org's GHA runner pool, which would make sense if there was a corresponding very large benefit.
> 
> I think it already uses a runner per job, so we are already using 6 runners: https://docs.github.com/en/actions/using-workflows/about-workflows <-- See 2 under Workflow basics

Oh, true. So this is a definite improvement.

**Review by zrhoffman — CHANGES_REQUESTED** [review]

File: .github/workflows/cache-config.unit.tests.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      - go.mod
+      - go.sum
+      - GO_VERSION
+      - grove/**.go
+      - lib/**.go
+      - traffic_monitor/**.go
+      - cache-config/**.go
+      - traffic_ops/**.go
+      - traffic_stats/**.go
+      - vendor/**.go
```
[the remark this answers is not in the document]
Comment (ericholguin) [reply .github/workflows/cache-config.unit.tests.yml:38]: Or no exclusions and always run unit tests on any go file changes?

[the remark this answers is not in the document]
Comment (zrhoffman) [reply .github/workflows/cache-config.unit.tests.yml:38]: Sure, that would work too

**Review by zrhoffman — APPROVED** [review]
Looks good, thanks!

**Comment by zrhoffman** [comment]
> One thing Zach didn't mention: according to what I'm seeing here, the lib tests only run if code in the lib changes. But that library is linked to from, I think, every other Go component. So when lib changes, we need to run ALL of those tests, not just the lib tests. Also, things that depend on the Go client need to be re-tested when the Go client changes. I'm actually not sure that we're doing that today - but if not it's a gap in testing that should be closed.

Looks like this is addressed, would you mind re-reviewing (and merging, if it's ready) @ocket8888?

**Review by ocket8888 — APPROVED** [review]

**Pull request merged**